[CDAP-938] partitioned file set api

anew opened the pull request:
This defines the interface PartitionedFileSet, plus some supporting classes:

- Partitioning to declare the partitioning schema of a dataset
- PartitionKey to specify the partition keys of a partition, used when adding partitions to a dataset
- PartitionFilter to specify a condition over the partitions, when reading from a dataset.
- PartitionedFileSetProperties to specify the partitioning in dataset properties along with other file set properties
- PartitionedFileSetArguments to specify dataset arguments (such as keys or filters) at runtime 

Also adds the test cases for these supporting classes.
 
Most of this is builders, getters, setters, etc., but there is some trickiness in the dealing with heterogenous types in a partition key or filter (conditions over Strings and Integers in a single map).

Next PR will contain the implementation of PartitionedFileSet.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionFilter.java:
```diff
@@ -0,0 +1,245 @@
+      try {
+        // if lower and upper are identical, then this represents an equality condition.
+        if (isSingleValue()) {
+          return getValue().equals(value);
```
Since you require `value` as `Comparable, would it be better to compare with `getValue().compareTo(value) == 0`?

anew replied:
sure that would work, too.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionFilter.java:
```diff
@@ -0,0 +1,245 @@
+   * @return the individual conditions of this filter.
+   */
+  public Map<String, Condition<? extends Comparable>> getConditions() {
+    return ImmutableMap.copyOf(conditions);
```
Maybe better to do a copy once in the constructor.

anew replied:
Oh yeah. I was planning to do that but forgot.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionFilter.java:
```diff
@@ -0,0 +1,245 @@
+
+  @Override
+  public String toString() {
+    return conditions.values().toString();
```
Just wonder if it's useful to have the keys as well?

anew replied:
The values are Conditions and a condition also has the field name. So this would be redundant. Oh, but perhaps I have to add that to Condition.toString()

anew replied:
Actually it already has the fieldName.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionFilter.java:
```diff
@@ -0,0 +1,245 @@
+                                                               @Nullable T lower,
+                                                               @Nullable T upper) {
+      Preconditions.checkArgument(field != null && !field.isEmpty(), "field name cannot be null or empty.");
+      Preconditions.checkArgument(lower == null || !lower.equals(upper), "Unsatisfiable condition: " +
```
Should it be using the `compareTo` instead of `equals`? `Comparable` only recommends to have `equals` behaves the same as `compareTo`, but not required.

anew replied:
sure, wiil do.

anew replied:
Actually after discussing in person, it seems more reasonable to allow these conditions. The upper and lower bound may be programmatic parameters in some code that generates filter. It is valid to pass a condition that can be satisfied in general, but can't under current context.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionKey.java:
```diff
@@ -0,0 +1,156 @@
+  /**
+   * @return all field names and their values in a map.
+   */
+  public Map<String, ? extends Comparable> getFields() {
```
Should the return type be just `Map<String, Comparable>`?

anew replied:
yes

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/Partitioning.java:
```diff
@@ -0,0 +1,215 @@
+    };
+
+    public abstract Comparable parse(String value);
+    public abstract <T> byte[] toBytes(T value);
```
Should be `<T extends FieldType>`

anew replied:
no. If FieldType is STRING, then T should be String. I am not super happy with this class, this is where you hit the limits of the Java type system, and generics only get you that far.

chtyim replied:
I see. It's actual value. Then should it be just `Object` (like the `validateType` method)?

anew replied:
Yes, but making it require Comparable protects against some type errors at compile time. I actually changed validateType to require a comparable.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionedFileSet.java:
```diff
@@ -0,0 +1,84 @@
+  /**
+   * @return the underlying (embedded) file set.
+   */
+  public FileSet getUnderlyingFileSet();
```
Should we call it `getEmbeddedFileSet` then?

anew replied:
We could. I named it the same as in TimePartitionedFileSet. Then we change both.

anew replied:
Oh if I change that in TPFS, that will break the stream adapter. Not sure if I want to do that now.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionedFileSetArguments.java:
```diff
@@ -0,0 +1,153 @@
+      String stringValue = keyArguments.get(fieldName);
+      Comparable fieldValue = convertFieldValue("key", "value", fieldName, fieldType, stringValue, false);
+      @SuppressWarnings({ "unchecked", "unused" }) // we know it's type safe, but Java does not
+      PartitionKey.Builder unused = builder.addField(fieldName, fieldValue);
```
Why need to have the `unused` field?

anew replied:
I can only suppress warnings on a declaration, hence using a temp variable. Then I have to add "unused" to the suppressed warnings, otherwise I get a warning about an unused. Java sometimes makes it hard to make your code free of warnings :)

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionedFileSetArguments.java:
```diff
@@ -0,0 +1,153 @@
+      Comparable fieldValue = convertFieldValue("filter", "value", fieldName, fieldType, stringValue, true);
+      if (null != fieldValue) {
+        @SuppressWarnings({ "unchecked", "unused" }) // we know it's type safe, but Java does not
+        PartitionFilter.Builder unused = builder.addValueCondition(fieldName, fieldValue);
```
Same here, I don't think the `unused` field is necessary.

anew replied:
it is, otherwise I get a warning about an unused variable

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionedFileSetProperties.java:
```diff
@@ -0,0 +1,99 @@
+   *
+   * @param properties the dataset properties
+   */
+  public static Partitioning getPartitioning(Map<String, String> properties) {
```
Looks like this method can return `null` for `Partitioning`. It's better to annotate this method with `@Nullable` especially it's cdap-api.

chtyim replied:
A followup question is, should we use `null` or an empty `Partitioning` to represent no partitioning is support ?

anew replied:
Agreed on the annotation and also adding that to javadocs. I still think we want to return null here. It indicates that no partitioning was specified in the properties. That is different from an empty partitioning (which, by the way, is not allowed).

anew commented:
@chtyim I have moved those FieldType methods (except for parse() which is needed by PartitionedFileSetProperties) into a class outside of the API, also addressed your other comments. Please take another look

chtyim commented:
LGTM
